### PR TITLE
Avoid allocations due to use of Keys/Values instead of GetEnumerator

### PR DESCRIPTION
**Pull request by sharwell**
`ImmutableDictionary<TKey, TValue>` provides a non-allocating implementation of `GetEnumerator()`, but does not offer non-allocating implementations of the `Keys` or `Values` properties (dotnet/corefx#249 and linked issues). This pull request adds a `Deconstruct` extension method for `KeyValuePair<TKey, TValue>`, leveraging a new C# feature to provide almost identical use-site code for these properties without actually incurring the allocation overhead.

:memo: This pull request builds on the change in #19125 to avoid merge conflicts (~~the first commit in the branch overlaps~~) (edit: no longer applies).

## Ask Mode

**Customer scenario**

No observable changes.

* Use `GetEnumerator()` instead of `Keys` where applicable
* Use `GetEnumerator()` instead of `Values` where applicable
* Add a `Deconstruct` extension method so the above changes can be implemented with minimal changes to the actual code structure

**Bugs this fixes:**

N/A

**Workarounds, if any**

None needed.

**Risk**

Low (no behavior changes).

**Performance impact**

Performance is improved by eliminating unnecessary allocations.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

Missed during code review?

**How was the bug found?**

Internal code review.

### CODE REVIEW

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for review (let me know if anyone else should be added)

Please note the comment regarding the first commit overlapping with #19125.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs [hunk 1]
```diff
@@ -215,9 +215,9 @@ internal bool IsSupportedDiagnostic(DiagnosticAnalyzer analyzer, Diagnostic diag
                 var isSuppressed = !diag.IsEnabledByDefault;
 
                 // If the user said something about it, that overrides the author.
-                if (diagnosticOptions.ContainsKey(diag.Id))
+                if (diagnosticOptions.TryGetValue(diag.Id, out var severity))
                 {
-                    isSuppressed = diagnosticOptions[diag.Id] == ReportDiagnostic.Suppress;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs]: We really sohuld have an analyzer fro that pattern

Comment (sharwell) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs]: We really should.

**Comment by sharwell** [comment]
@dotnet-bot retest this please

**Comment by sharwell** [comment]
@dotnet-bot retest this please

**Comment by sharwell** [comment]
@MeiChin-Tsai for ask mode (at the request of @jaredpar)

**Comment by MeiChin-Tsai** [comment]
approved.

**Comment by jcouv** [comment]
Nice

**Comment by sharwell** [comment]
@jcouv I think this is my favorite use of C# 7 to date. It totally resolved something I previously found rather frustrating. :smile:

**Pull request merged**